### PR TITLE
[Token] Add token marketplace package in move-examples

### PR DESCRIPTION
**Pull request by areshand**
### Description
https://github.com/aptos-labs/aptos-core/issues/3067

### Test Plan
add unit test

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3276)

### CODE REVIEW

**Comment by ericEnjoy** [comment]
I am wonder if this PR will merge into next devnet? It's a breaking change and ecosystem need time to upgrade theri contracts

**Comment by areshand** [comment]
> I am wonder if this PR will merge into next devnet? It's a breaking change and ecosystem need time to upgrade theri contracts

won't be included in the coming devnet

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/bid.move:74]: not clear how the smart contract will track bids against a specific token

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/bid.move:1]: would be good to document the flow at the top here.

**Review by davidiw — COMMENTED** [review]
The code currently lacks a clear story. For example, it isn't clear to me how this is used to drive an auction, how one might build royalties, or add fees. I think if we want to go in this direction, maybe the better choice is to construct a pretty comprehensive market place derived from the work here and #1204 and create something exploratory in `move-examples` until we land on a solution that at least demonstrates a clear end-to-end. From there, we can decide do we want to be opinionated or break it down into composable pieces.

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:74]: > The code currently lacks a clear story. For example, it isn't clear to me how this is used to drive an auction, how one might build royalties, or add fees. I think if we want to go in this direction, maybe the better choice is to construct a pretty comprehensive market place derived from the work here and #1204 and create something exploratory in `move-examples` until we land on a solution that at least demonstrates a clear end-to-end. From there, we can decide do we want to be opinionated or break it down into composable pieces.

will add an example to this PR to show the e2e flow.

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:3]: add doc pointing to the example

**Comment by davidiw** [comment]
I'm happy to land if we move this all to move-examples. I think to move forward as is, I'd think we'd probably need to review this a lot closer and it would ideally be accompanied by a guide, tutorial, etc.

I'd rather land it first in move-examples and make sure it doesn't bit rot, but open to discussions.

**Comment by areshand** [comment]
> I'm happy to land if we move this all to move-examples. I think to move forward as is, I'd think we'd probably need to review this a lot closer and it would ideally be accompanied by a guide, tutorial, etc.
Yes, it is good suggestion. we can have more time to get more feedbacks on this work

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/bid.move:14]: nit ordering here

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/bid.move:96]: this `offer_price * token_amount` should probably be commented somewhere besides reading the code.

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/bid.move:124]: naming feels like it should be more like `cancel_bid` or just `withdraw_bid`.

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/bid.move:158]: This looks like something that should be defined in an auction_house.move and called `complete_auction`. It Also seems like a backdoor, where I can take any successful bid and complete it, even if it isn't the highest bid for an auction.

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/bid.move:241]: There's no point in using a guid here, we can just as well use a u64 counter. If we aren't sharing the GUID across users, then address portion is irrelevant and the u64 is globally unique for that user (bid resource + user account + u64).

File: aptos-move/framework/aptos-token/sources/listing.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/listing.move:21]: could this be both?

File: aptos-move/framework/aptos-token/sources/listing.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/listing.move:27]: maybe `create_listing` or just `list`. the term as used implies you're creating a list data structure, e.g., a vector.

Comment (davidiw) [reply aptos-move/framework/aptos-token/sources/listing.move:27]: why are we exposing this externally?

File: aptos-move/framework/aptos-token/sources/listing.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/listing.move:89]: why do we need this versus just the entry function or the list function.

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (davidiw) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:17]: nit ordering

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (davidiw) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:19]: whitespace weirdness

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (davidiw) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:102]: do we really care?

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (davidiw) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:154]: we need to figure out naming here and stick to a convention :P

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (davidiw) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:205]: probably should just be bid.

**Review by davidiw — COMMENTED** [review]
I read most of it. I think it is off to a great start. There's appears to be a security hole where any bid can be used to execute a listing and perhaps more than once?

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:124]: cancel_bid is defined in the auction_house, where the auction owner removes one bid from `bids` in `Auction`.  This removed bid won't participate in the bidding.

withdraw_fund_from_bid here is for the bidder to extract the coins from his bid without any approval from the Auction House

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:158]: This is a library function supposed to be called by an auction house. It is designed to be flexible to match a listing with a bid chosen the by auction house. 

To execute this function, it needs to have `Listing<CoinType>`, which is owned by auctioneer, who control the auction logic anyway. No anyone can execute this function.

File: aptos-move/framework/aptos-token/sources/listing.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/listing.move:21]: No, if instant_sale is true, the sale immediately happens if someone bid higher than the listed price. We don't need to wait till the end of the auction.

File: aptos-move/framework/aptos-token/sources/listing.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/listing.move:27]: This is for the use-case of launch-pad or large marketplace where they want to build their listing through partnering with some creators. The can store these listings in their deployed market_place contracts.

File: aptos-move/framework/aptos-token/sources/listing.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/listing.move:89]: This is for 2nd usecase where creator want to directly list under their account, own the listing sell them directly.

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:102]: 1. simply the auction logic so that new accepted bids are strictly higher than previous bidder.
2. participants won't be annoyed by someone just adding a very very small amount to win the auction in the last minute

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [reply aptos-move/framework/aptos-token/sources/bid.move:124]: I don't know what the purpose of this is but to effectively cancel my own bid. withdraw_fund(s) from bid doesn't necessarily imply you would remove the bid from the bidding table.

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (davidiw) [reply aptos-move/framework/aptos-token/sources/bid.move:158]: Point being that the auctioneer can cheat with the logic as it stands.

Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:158]: I don't think this is an issue that we should police. We provide a library and example. If they want to use the library and create any malicious module, let the them shoot on their own feet then?

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:124]: Yes, withdraw doesn't remove bid from listing. The bid library knows nothing about the auction house's data. 
The purpose is to allow bidder to extract their coin and store into their account after the auction ends without the auctioneer's approval.  renamed

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:96]: addressed

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:14]: addressed

File: aptos-move/framework/aptos-token/sources/bid.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/bid.move:241]: a global counter shared by multiple users will limit parallelism.

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:17]: addressed

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:19]: addressed

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:154]: updated

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:205]: that is already used

Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move]: updated

**Comment by areshand** [comment]
Add token_proxy to separate the experiment token modules from core token modules while retain friend relation with having significant change to product related features

When we want to deprecate existing bid/listing, we can abort the existing modules and switch the proxy to point to new modules

This is supposed to deploy to testnet for now and should only deploy to mainnet once we have confidence on the usability and security

File: aptos-move/framework/aptos-token/sources/token.move
Comment (davidiw) [inline aptos-move/framework/aptos-token/sources/token.move:619]: nice

**Review by davidiw — APPROVED** [review]

File: aptos-move/framework/aptos-token/sources/token.move
Comment (movekevin) [inline aptos-move/framework/aptos-token/sources/token.move:609]: Hmm what if someone issues two different withdraw capabilities? Wouldn't this lead to tug of war?

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:105]: Can this be combined with initialize_auction_house?

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:131]: What about withdraw_expiration_sec?

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:128]: Can we validate amount > 0 as well?

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:202]: nit: <ListingEvent> type argument is not needed.

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:217]: What happens if someone creates multiple auctions at the same time? Wouldn't cur_auction_id be wrong as it only tracks the latest created auction?

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:248]: Why is this check here and not in auction creation?

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:237]: There might be some racing condition here if multiple people try to submit bids in the same block. The subsequent bids might always fail.

File: aptos-move/move-examples/nft_auction_house/sources/bid.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/bid.move:118]: This multiplication can overflow.

File: aptos-move/move-examples/nft_auction_house/sources/bid.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/bid.move:107]: What happens when someone bids multiple times? For example I bid 5 coins and then 15 coins later on. Wouldn't this take 20 coins out of my account? If so, it doesn't seem correct to me as my later higher bids should override.

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (movekevin) [inline aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:305]: It seems strange and can open up edge cases to let highest bidder cancel their bid. This doesn't seem typical in the standard auctions?

File: aptos-move/framework/aptos-token/sources/token.move
Comment (areshand) [reply aptos-move/framework/aptos-token/sources/token.move:609]: no war only first come first serve if token is limited

Comment (movekevin) [reply aptos-move/framework/aptos-token/sources/token.move:609]: I see - I missed the token_owner field. With that said though, in general, is it safe to allow issuing multiple WithdrawCapability objects?

Comment (movekevin) [reply aptos-move/framework/aptos-token/sources/token.move:609]: There are edge cases for example on amount as well. Owner can also use WithdrawCap to double or triple sell their tokens

Comment (areshand) [reply aptos-move/framework/aptos-token/sources/token.move:609]: It is common that people listing their token on multiple places for sale.

File: aptos-move/move-examples/nft_auction_house/sources/bid.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/bid.move:107]: this would just record a bid with a bid_id.  This mainly tries to provide a library for bidding. The actual auction logic can be found in the auction_house example. The auction_house will choose the highest bid to complete the auction

File: aptos-move/move-examples/nft_auction_house/sources/bid.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/bid.move:118]: yes, a overflow error will happen. feel the error is already very clear enough

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:105]: no, we don't want to call the initialize_auction_house for each new coinType

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:131]: that is checked in `create_list` to ensure withdraw_expiration_sec larger than expiration_sec

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:128]: added

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:217]: This is for generating auction_id by tracking the existing largest auction_id.

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:248]: Good suggestion, moved this check to auction creation

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:237]: The bid TXN should be executed in sequence for this auction.

File: aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move
Comment (areshand) [reply aptos-move/move-examples/nft_auction_house/sources/auction_house_own_listing.move:305]: This is product feature. For example, opensea allows you to cancel bid while magic eden doesn't. But generally cancel bid cost gas. I will also collect feedbacks from our devs on this as well

**Comment by areshand** [comment]
Hi @clay-aptos , I added a readme to this new package. Could you please review it ?

**Review by clay-aptos — APPROVED** [review]
Thanks, Bo. This LGTM with some light edits I made directly in a commit here.

**Pull request merged**
